packages/compiler/core/checker.ts

```diff
@@ -4586,21 +4582,22 @@ export function getEffectiveModelType(
       continue;
     }
 
-    // Add any derived types we observe to both sides. A derived type can
-    // substitute for a base type in these sets because derived types have
-    // all the properties of their bases.
+    // Add derived sources as we encounter them. If a model is sourced from
+    // a base property, then it can also be sourced from a derived model.
+    //
+    // (Unless it is overridden, but then the presence of the overridden
+    // property will still cause the the base model to be excluded from the
+    // candidates.)
     //
-    // NOTE: Once property overrides are allowed, this code will need to
-    // be updated to check that the current property is not overridden by
-    // the derived type before adding it here. An override would invalidate
-    // this substitution.
+    // Note: We depend on the order of that spread and intersect source
+    // properties here, which is that we see properties sourced from derived
+    // types before properties sourced from their base types.
     addDerivedModels(sources, candidates);
-    addDerivedModels(candidates, sources);
```

Comment:
The second call was unnecessary due to the ordering which is also needed to handle overrides correctly. Turns out the only thing wrong was countInheritedProperties, but I added a test and reworded the comment that thought we'd need more changes once overriding happened.